include/seastar/core/weak_ptr.hh

```diff
@@ -139,6 +139,12 @@ public:
         _ptr_list.push_back(ptr);
         return ptr;
     }
+
+    weak_ptr<const T> weak_from_this() const noexcept {
+        weak_ptr<const T> ptr(static_cast<const T*>(this));
+        _ptr_list.push_back(reinterpret_cast<weak_ptr<T>&>(ptr));
```

Comment:
Instead of this problematic construct (as you saw, the compiler didn't like it either), I wonder if you can make the _ptr_list definition use something like `const T` or `std::remove_const<T>` instead of T, which will maybe allow you to push the weak_ptr<const T> onto _ptr_list.
(I haven't thought this through so maybe you can come up with a better solution)

Reply:
I thought about it too, but couldn't break the boost intrusive list requirement to explicitly specify the type of objects it keeps :( I pushed an updated version that made the compiler happy